sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ChannelReaderExtensions.cs

```diff
@@ -0,0 +1,122 @@
+                                                                    [EnumeratorCancellation]CancellationToken cancellationToken)
+        {
+            Guard.ArgumentNotNull(nameof(reader), reader);
+
+            if (maximumWaitTime.HasValue)
+            {
+                Guard.ArgumentNotNegative(nameof(maximumWaitTime), maximumWaitTime.Value);
+            }
+
+            T result;
+
+            var waitToken = cancellationToken;
+            var waitSource = default(CancellationTokenSource);
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    if (reader.TryRead(out result))
+                    {
+                        waitSource?.CancelAfter(maximumCancellationDelay);
+                        yield return result;
+                    }
+                    else if (reader.Completion.IsCompleted)
+                    {
+                        // If the channel was marked as final, then await the completion task to surface any exceptions.
+
+                        await reader.Completion.ConfigureAwait(false);
+                        break;
+                    }
+                    else
+                    {
+                        try
+                        {
+                            if (maximumWaitTime.HasValue)
+                            {
+                                if ((waitSource == null) || (waitSource.IsCancellationRequested))
+                                {
+                                    waitSource?.Dispose();
+                                    waitSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                                }
+
+                                waitSource.CancelAfter(maximumWaitTime.Value);
+                                waitToken = waitSource.Token;
+                            }
+
+                            // Wait for an item to be available in the channel; if it becomes so, then
+                            // reset the loop so that it can be read and emitted.
+
+                            if (await reader.WaitToReadAsync(waitToken).ConfigureAwait(false))
+                            {
+                                waitSource?.CancelAfter(maximumCancellationDelay);
+                                continue;
+                            }
+                        }
+                        catch (OperationCanceledException)
+                        {
+                            // This is thrown when the wait token expires.  It may be caused by the maximum wait time
+                            // being exceeded or the main cancellation token being set.  Ignore this as an expected
+                            // case; if the iteration was canceled, it will be detected in the body of the loop and
+                            // appropriate action taken.
+
+                            waitSource?.Dispose();
```

Comment:
we have the same dispose in finally, any reason to have two?

Reply:
If we hit this `catch` block the token source has been canceled, either because the wait time expired and we have to emit a default item or because the main cancellation token was triggered.   In the case that we're emitting a default, the token source has to be recreated on the next tick and we don't hit the `finally`,  so I'm disposing here.   If we don't, any time the wait time is exceeded and we emit a default, we'll be throwing away a token source without disposing.

The finally is intended as a guard for cancellation via token or the channel being marked final occurring in a part of the loop where we aren't block on `WaitToReadAsync`.   

In the one case that the main token is cancelled while we are in `WaitToReadAsync`, we will hit both paths for dispose, but at that point `waitSource` is null anyway (by coincidence) to it's only being invoked once.   

Unless I'm overlooking something, I believe that both are needed.